pool.go

```diff
@@ -0,0 +1,212 @@
+// once a certain number of modules have been fetched.
+type Pool struct {
+	mutex           *sync.Mutex
+	runtimes        []*RuntimeTracker
+	returnedModules map[uintptr]uint64
+	nextId          uint64
+	maxFetch        uint64
+}
+
+// NewPool creates a new *Pool.
+func NewPool() *Pool {
+	r, compiled := createRuntime(context.Background())
+	pool := &Pool{
+		mutex: &sync.Mutex{},
+		runtimes: []*RuntimeTracker{{
+			id:       1,
+			r:        r,
+			compiled: compiled,
+			modules:  make([]api.Module, 0, 16),
+			max:      0,
+			fetches:  0,
+		}},
+		returnedModules: make(map[uintptr]uint64),
+		nextId:          2,
+		maxFetch:        128,
+	}
+	return pool
+}
+
+// Get returns a new module from the pool.
+func (pool *Pool) Get() api.Module {
+	pool.mutex.Lock()
+	defer pool.mutex.Unlock()
+
+	ctx := context.Background()
+	rtracker := pool.runtimes[len(pool.runtimes)-1]
+	rtracker.fetches++
+	// If we've used up the number of fetches allowed in this runtime, then we'll create a new one
+	if rtracker.fetches >= pool.maxFetch {
+		r, compiled := createRuntime(ctx)
+		rtracker = &RuntimeTracker{
+			id:       pool.nextId,
+			r:        r,
+			compiled: compiled,
+			modules:  make([]api.Module, 0, 16),
+			max:      0,
+			fetches:  0,
+		}
+		pool.runtimes = append(pool.runtimes, rtracker)
+		pool.nextId++
+	}
+	var module api.Module
+	// If the runtime has no modules remaining, then we need to create a new module
+	if len(rtracker.modules) == 0 {
+		rtracker.max++
+		var err error
+		module, err = rtracker.r.InstantiateModule(ctx, rtracker.compiled, icuConfig)
+		if err != nil {
+			panic(err)
+		}
+	} else {
+		// Pop the last module from the slice
+		module = rtracker.modules[len(rtracker.modules)-1]
+		rtracker.modules = rtracker.modules[:len(rtracker.modules)-1]
+	}
+	// Now we need to track that this module is being returned
+	pool.returnedModules[reflect.ValueOf(module).Pointer()] = rtracker.id
+	runtime.SetFinalizer(module, func(module api.Module) {
+		pool.finalized(module)
+	})
+	return module
+}
+
+// Put returns the module to the pool.
+func (pool *Pool) Put(module api.Module) {
+	pool.mutex.Lock()
+	defer pool.mutex.Unlock()
+	pool.receivedModule(module, true)
+}
+
+// finalized is called by the finalizer, and only exists to catch orphaned modules.
+func (pool *Pool) finalized(module api.Module) {
+	pool.mutex.Lock()
+	defer pool.mutex.Unlock()
+	pool.receivedModule(module, false)
+}
+
+// receivedModule is called when either the module is returned through Put, or the finalizer catches an orphaned module
+// through finalized.
+func (pool *Pool) receivedModule(module api.Module, isPut bool) {
+	// Remove the finalizer that was set when the object was fetched.
+	// This is only called from Put, as the finalizer is being called so we don't want to remove it.
+	if isPut {
+		runtime.SetFinalizer(module, nil)
+	}
+	// Grab the runtime ID and remove the module from the tracking map
+	ptr := reflect.ValueOf(module).Pointer()
```

Comment:
If this is slow, it might make sense to get it a single time at creation, and for the pool to vend an object which includes both the `api.Module` and its `uintptr`

Reply:
This isn't the slow portion. It's `wazero.NewRuntime(ctx)` that's slow (like half a second on my machine). That's why we can't just create a new one on every request. Second to that is `InstantiateModule`. Everything else is essentially instant in comparison. The slowdown is due to having to periodically a new runtime, not really anything else.